spark/src/main/scala/geotrellis/spark/clip/ClipToGrid.scala

```diff
@@ -113,29 +113,32 @@ object ClipToGrid {
     clipFeature: (Extent, Feature[G, D], Predicates) => Option[Feature[Geometry, D]],
     feature: Feature[G, D]
   ): Iterator[(SpatialKey, Feature[Geometry, D])] = {
-    val pointPredicates =
+
+    /* Unique keys that this Feature touches */
+    val keys: Set[SpatialKey] = mapTransform.keysForGeometry(feature.geom)
+
+    lazy val mpOrLinePredicates =
       new Predicates {
         def covers(e: Extent) = false
-        def coveredBy(e: Extent) = true
+        def coveredBy(e: Extent) = keys.size < 2
       }
 
-    val mpOrLinePredicates =
+    def preparedPredicates(pg: PreparedGeometry) =
       new Predicates {
-        def covers(e: Extent) = false
-        def coveredBy(e: Extent) =
-          feature.geom.jtsGeom.coveredBy(e.toPolygon.jtsGeom)
+        def covers(e: Extent) = pg.covers(e.toPolygon.jtsGeom)
+        def coveredBy(e: Extent) = keys.size < 2
       }
 
-    def polyPredicates(pg: PreparedGeometry) =
+    lazy val polyPredicates =
       new Predicates {
-        def covers(e: Extent) = pg.covers(e.toPolygon.jtsGeom)
-        def coveredBy(e: Extent) = pg.coveredBy(e.toPolygon.jtsGeom)
+        def covers(e: Extent) = feature.geom.jtsGeom.covers(e.toPolygon.jtsGeom)
+        def coveredBy(e: Extent) = keys.size < 2
       }
 
-    def gcPredicates =
+    lazy val gcPredicates =
       new Predicates {
         def covers(e: Extent) = feature.geom.jtsGeom.covers(e.toPolygon.jtsGeom)
-        def coveredBy(e: Extent) = feature.geom.jtsGeom.coveredBy(e.toPolygon.jtsGeom)
+        def coveredBy(e: Extent) = keys.size < 2
```

Comment:
The `coveredBy` condition is now the same for all geometry types. It could be factored out with the `keys.size < 2` check happening once below in the `val iterator = ...`. This would avoid billions of pattern matches on the full data set.